doc/source/development/rfc/rfc86_column_oriented_api.rst

```diff
@@ -0,0 +1,692 @@
+or using ResetReading()/GetNextFeature() while using a ArrowArrayStream is
+strongly discouraged and may lead to unexpected results. As a rule of thumb,
+no OGRLayer methods that affect the state of a layer should be called on a
+layer, while an ArrowArrayStream on it is active.
+
+A potential usage can be:
+
+.. code-block:: cpp
+
+    struct ArrowArrayStream stream;
+    if( !poLayer->GetArrowStream(&stream, nullptr))
+    {
+        fprintf(stderr, "GetArrowStream() failed\n");
+        exit(1);
+    }
+    struct ArrowSchema schema;
+    if( stream.get_schema(&stream, &schema) == 0 )
+    {
+        // Do something useful
+        schema.release(schema);
+    }
+    while( true )
+    {
+        struct ArrowArray array;
+        // Look for an error (get_next() returning a non-zero code), or
+        // end of iteration (array.release == nullptr)
+        //
+        if( stream.get_next(&stream, &array) != 0 ||
+            array.release == nullptr )
+        {
+            break;
+        }
+        // Do something useful
+        array.release(&array);
+    }
+    stream.release(&stream);
+
+The papszOptions that may be provided is a NULL terminated list of key=value
+strings, that may be driver specific.
+
+OGRLayer has a base implementation of GetArrowStream() that is such:
+
+- The get_schema() callback returns a schema whose top-level object returned is
+  of type Struct, and whose children consist in the FID column, all OGR attribute
+  fields and geometry fields to Arrow fields.
+  The FID column may be omitted by providing the INCLUDE_FID=NO option.
+
+  When get_schema() returns 0, and the schema is no longer needed, it must
+  be released with the following procedure, to take into account that it might
+  have been released by other code, as documented in the Arrow C data
+  interface:
+
+  .. code-block:: c
+
+          if( out_schema->release )
+              out_schema->release(out_schema)
+
+
+- The get_next() callback retrieve the next record batch over the layer.
+
+  out_array is a pointer to a ArrowArray structure, that can be in a uninitialized
+  state (the method will ignore any initial content).
+
+  The default implementation uses GetNextFeature() internally to retrieve batches
+  of up to 65,536 features (configurable with the MAX_FEATURES_IN_BATCH=num option).
+  The starting address of buffers allocated by the
+  default implementation is aligned on 64-byte boundaries.
+
+  The default implementation outputs geometries as WKB in a binary field,
+  whose corresponding entry in the schema is marked with the metadata item
+  ``ARROW:extension:name`` set to ``WKB``. Specialized implementations may output
+  by default other formats (particularly the Arrow driver that can return geometries
+  encoded according to the GeoArrow specification (using a list of coordinates).
+  The GEOMETRY_ENCODING=WKB option can be passed to force the use of WKB (through
+  the default implementation)
```

Comment:
One thing I'm not super-clear on here -- if the requested geometry format matches the underlying dataset's geometry format, does that mean there'll be absolutely **no** parsing/re-interpretation of the geometry on GDAL's side?

Reply:
> does that mean there'll be absolutely no parsing/re-interpretation of the geometry on GDAL's side?

yes, this is typically the case for GeoParquet that uses WKB as the storage format. And also for GeoPackage where only the GeoPackage geometry header is stripped from the SQlite blob to get a WKB geometry. No OGRGeometry object is instanciated then.

Reply:
I should precise that when the default implementation of OGRLayer is used, as GetNextFeature() is used underneath and returns a OGRGeometry, if the underlying format would happen to use WKB for storage, you'd have:  WKB --> OGRGeometry (in GetNextFeature() -> WKB (in OGRLayer::GetNextArrowArray()). That would for example be the case for the PostGIS driver currently (with EWKB as the source encoding)

Reply:
(I realise this is off-topic, sorry!)

> yes, this is typically the case for GeoParquet that uses WKB as the storage format. And also for GeoPackage where only the GeoPackage geometry header is stripped from the SQlite blob to get a WKB geometry.

What's your thoughts regarding having a similar short-cut approach for the non-column oriented apis? I.e. being able to read features from a gpkg where the geometry is left as just the wkb binary content, and no interpretation is made on gdal's side. In the gpkg case this could be a good optimisation opportunity for some clients, in that they could request the wkb content directly if they are going to be constructing their own geometry representation of the features (ie gpkg -> wkb -> client geometry classes ) and want to avoid the extra representation (gpkg -> wkb -> ogr -> client geometry )

Reply:
That's a good idea, but I'm not sure how to implement this in a generic option. I guess this could be an open option implemented in the drivers where that would make sense that would turn the geometry column as a regular Binary field. I'm not sure how to inform client of which column(s) has the geometry. maybe some new field in the OGRGeomFieldDefn class.